### PR TITLE
[3.0] Improve testing of MixKhronosData and PrettifyNames mods

### PR DESCRIPTION
**Pull request by Exanite**
# Summary of the PR

(Description is heavily outdated)

This PR does the following:

- Replace the `OverzealousNameTrimming` test case with 3 new test cases
   - The old test case was failing and did not reflect the intended behavior of the codebase.
   - The new test cases test the intended behavior in a more granular manner.
   - The new test cases are:
      - `PrettifyNames_TrimsSharedPrefix`
      - `PrettifyNames_TrimsSharedPrefix_AfterRemovalOf_VendorSuffixes`
      - `MixKhronosData_IdentifiesVendorSuffixes`
- Add a few utilities to make testing easier
- Add tests for some of the features I added to MixKhronosData
- Ensure that the new test cases pass (because they did not)
   - Notably:
      - Prefix identification (NameTrimmer) now runs after affixes are removed
      - NameTrimmer now trims based off of the primary names (from the viewpoint of NameTrimmer, these are the names with affixes removed) instead of the original names (these are the managed names that existed in the codebase at the start of PrettifyNames)

Note: I used underscores for the test names unlike the existing tests since it was really hard to clearly describe the edge case conditions without them.

Todo list:

- [x] Fixed: `OcclusionQueryParameterNameNV` is currently being trimmed as `QueryParameterNameNV` when it is the only type being trimmed
   - I'm assuming this is unwanted behavior 
   - It looks like `NameTrimmer` just trims the first word in this case
   - Note that I also provide `gl` as a prefix hint. This might be related. I need to investigate.
- [ ] Fix failing test cases
- [x] When do we handle global prefix identification (`NameTrimmer`)? Before stripping affixes or after?
   - This affects what part of the name is visible to `NameTrimmer`.
   - Note that most `NameAffix`es will be suffixes.
   - If we run `NameTrimmer` after stripping affixes, the effect is less of the name is trimmed, which should fix a few cases where `NameTrimmer` trims too much, which can cause affix stripping to output `""` or `"_"`, which sub-sequentially errors during prettification.
   - For reference, the current process is:
      - Trim prefixes (`NameTrimmer`)
      - Strip affixes
      - Prettify
      - Add affixes back according to PrettifyNames' config
   - Been thinking about this for a while and my conclusion is that we do want to strip affixes first. Will keep this section here for reference.
- [x] Consider removing `OverzealousNameTrimmingFixupIsNotOverzealousForOpenAL` test case. It overlaps with my new tests, but doesn't hurt. The name might be misleading though. 

# Related issues, Discord discussions, or proposals
Discord discussion on `OverzealousNameTrimming`: https://discord.com/channels/521092042781229087/587346162802229298/1458562017475432480

# Further Comments
N/A

### CODE REVIEW

File: sources/SilkTouch/SilkTouch/Mods/MixKhronosData.cs [hunk 1]
```diff
@@ -356,8 +356,8 @@ .. currentConfig.NonStandardExtensionNomenclature
     /// <inheritdoc />
     public async Task ExecuteAsync(IModContext ctx, CancellationToken ct = default)
     {
-        var currentConfig = cfg.Get(ctx.JobKey);
         var jobData = Jobs[ctx.JobKey];
+        var currentConfig = jobData.Configuration;
```
Comment (Exanite) [inline sources/SilkTouch/SilkTouch/Mods/MixKhronosData.cs:360]: This is probably more correct and allows the config to only need to be specified in one place in MixKhronosDataTests.

File: sources/SilkTouch/SilkTouch/Naming/NameTrimmer.cs [hunk 2]
```diff
@@ -46,17 +46,12 @@ public void Trim(NameTrimmerContext context)
         {
             for (var i = 0; i < nPasses; i++) // try with both trimming name and non trimming name
             {
-                if (context.Names is null)
```
Comment (Exanite) [inline sources/SilkTouch/SilkTouch/Naming/NameTrimmer.cs:49]: No longer nullable

File: sources/SilkTouch/SilkTouch/Mods/PrettifyNames.cs [hunk 3]
```diff
@@ -47,7 +47,7 @@ IEnumerable<IJobDependency<INameTrimmer>> trimmerProviders
         /// The maximum length of an all capitals string to be treated as a single acronym, rather than as an all
         /// capitals word.
         /// </summary>
-        public int? LongAcronymThreshold { get; init; }
+        public int LongAcronymThreshold { get; init; } = 3; // TODO: Change default to 2 in next PR to match framework design guidelines
```
Comment (Exanite) [inline sources/SilkTouch/SilkTouch/Mods/PrettifyNames.cs:50]: General cleanup. This is easier to understand.

File: sources/SilkTouch/SilkTouch/Naming/CandidateNamesWithOriginal.cs [hunk 4]
```diff
@@ -1,24 +0,0 @@
...
-namespace Silk.NET.SilkTouch.Naming;
-
-/// <summary>
-/// Represents the set of primary and secondary options for the trimmed version of a name.
-/// Also includes the original, unmodified name.
-/// </summary>
-/// <param name="Primary">The preferred version of the trimmed name.</param>
-/// <param name="Secondary">The fallback versions of the trimmed name in case the primary does not work.</param>
-/// <param name="Original">The original, unmodified name.</param>
-public record struct CandidateNamesWithOriginal(
```
Comment (Exanite) [inline sources/SilkTouch/SilkTouch/Naming/CandidateNamesWithOriginal.cs:13]: This was only used by NameTrimmer so it has been moved there.

File: sources/SilkTouch/SilkTouch/Naming/NameTrimmer.cs [hunk 5]
```diff
@@ -35,7 +35,7 @@ public class NameTrimmer : INameTrimmer
     public void Trim(NameTrimmerContext context)
     {
         string? identifiedPrefix = null;
-        Dictionary<string, CandidateNamesWithOriginal> localNames = null!;
```
Comment (Exanite) [inline sources/SilkTouch/SilkTouch/Naming/NameTrimmer.cs:38]: This was used as a glorified list of tuples so it is a list of tuples now. The dictionary was never used for lookup purposes. If it is needed as a dictionary later, use the original name as the key instead of the trimming name (this was causing name collision issues).

File: sources/SilkTouch/SilkTouch/Naming/NameTrimmer.cs [hunk 6]
```diff
@@ -46,17 +46,12 @@ public void Trim(NameTrimmerContext context)
...
-                {
-                    continue;
-                }
-
                 // Attempt to identify the hint being used.
                 string? hint = null;
                 foreach (var candidateHint in context.Configuration.GlobalPrefixHints ?? [])
                 {
                     var match = true;
-                    foreach (var name in context.Names.Keys)
```
Comment (Exanite) [inline sources/SilkTouch/SilkTouch/Naming/NameTrimmer.cs:59]: This was trimming based off the original name instead of the primary (current) names. This means that NameTrimmer was ignoring modifications if any trimmer came before it.

File: sources/SilkTouch/SilkTouch/Naming/NameTrimmer.cs [hunk 7]
```diff
@@ -193,21 +188,23 @@ hint is null
...
+    /// A local names list is also returned.
+    /// This is the list of names to be used for the remainder of the trimming process
+    /// and contains the trimming name and original name.
     /// </returns>
-    protected (string Prefix, Dictionary<string, CandidateNamesWithOriginal>)? GetPrefix(
+    protected (string Prefix, List<AugmentedCandidateNames>)? GetPrefix(
         string? container,
         string? hint,
-        Dictionary<string, CandidateNames>? names,
+        Dictionary<string, CandidateNames> names,
```
Comment (Exanite) [inline sources/SilkTouch/SilkTouch/Naming/NameTrimmer.cs:199]: This is no longer nullable.

File: sources/SilkTouch/SilkTouch/Naming/NameTrimmer.cs [hunk 8]
```diff
@@ -234,42 +238,68 @@ bool naive
         // then it will trim ThingsRGB to sRGB and ThingRGB to RGB
         // a case like this is simple to add a special case for in the generator to handle sRGB specially,
         // but see ImageChannelOrder from spirv.h for a more problematic occurrence.
-        var prefix =
+        string prefix;
```
Comment (Exanite) [inline sources/SilkTouch/SilkTouch/Naming/NameTrimmer.cs:241]: Big diff, but only two changes:
1. The nested ternary was replaced with if statements. Ternaries are horrible to step through using a debugger.
2. I added the "One name. Can't determine prefix." case to fix `DoesNotTrimTypeName_WhenNotMatchingHint_AndOnlyOneType` (`OcclusionQueryParameterNameNV` was being trimmed as `QueryParameterNameNV` ONLY during test cases since the test cases usually only have one type)

File: sources/SilkTouch/SilkTouch/Naming/NameTrimmer.cs [hunk 9]
```diff
@@ -327,8 +357,32 @@ public virtual string GetTrimmingName(
...
 
-        return hint is not null && name.StartsWith(hint, StringComparison.OrdinalIgnoreCase)
-            ? $"{hint}_{name[hint.Length..].Trim('_').LenientUnderscore()}"
-            : name.LenientUnderscore();
+        if (hint is not null && name.StartsWith(hint, StringComparison.OrdinalIgnoreCase))
+        {
+            return $"{hint}_{name[hint.Length..].Trim('_').LenientUnderscore()}";
+        }
+
+        return name.Trim('_').LenientUnderscore();
```
Comment (Exanite) [inline sources/SilkTouch/SilkTouch/Naming/NameTrimmer.cs:365]: Replaced ternary with ifs.
Added trim to what was originally the else branch.

File: sources/SilkTouch/SilkTouch/Naming/NameUtils.cs [hunk 10]
```diff
@@ -59,7 +59,8 @@ public static string Prettify(
 
         var ret = string.Join(
             null,
-            str.LenientUnderscore()
+            str.Trim('_')
+                .LenientUnderscore()
                 .Humanize()
```
Comment (Exanite) [inline sources/SilkTouch/SilkTouch/Naming/NameUtils.cs]: Humanize treats `RGB32F` and `RGB32F_` differently. It breaks the first up into 3 words. It does not break the second one at all.

File: sources/SilkTouch/SilkTouch/Naming/TrimmerOrder.cs [hunk 11]
```diff
@@ -8,11 +8,11 @@ namespace Silk.NET.SilkTouch.Naming;
 /// </summary>
 internal enum TrimmerOrder
 {
-    NameTrimmer = 100,
-    NameTrimmer217 = 110,
-    NameTrimmer218 = 120,
+    NameAffixerEarlyTrimmer = 100,
```
Comment (Exanite) [inline sources/SilkTouch/SilkTouch/Naming/TrimmerOrder.cs:11]: This makes it so that the NameTrimmer does not see any declared affixes (currently all suffixes).
This ensures that prefixes, base names, and suffixes never overlap.

**Review by Exanite — COMMENTED** [review]
Added notes on why changes were made.
Will self review the test outputs tomorrow.
